subprojects/core/src/test/groovy/org/gradle/execution/plan/ExecutionNodeAccessHierarchyTest.groovy

```diff
@@ -172,6 +173,31 @@ class ExecutionNodeAccessHierarchyTest extends Specification {
         assertNodesAccessing("/other", root)
     }
 
+    def "can query for nodes accessing using excludes for parts of paths"() {
+        def childNode = Mock(Node)
+
+        hierarchy.recordNodeAccessingLocations(childNode, ["/some/root/child/relative/path"])
+        expect:
+        hierarchy.getNodesAccessing("/some/root", excludes("child")).empty
+        hierarchy.getNodesAccessing("/some/root", includes("child")).empty
+        hierarchy.getNodesAccessing("/some/root", includes("child/**")) == ImmutableSet.of(childNode)
+    }
+
+    def "can record file trees with excludes for parts of paths"() {
+        def excluding = Mock(Node)
+        def including = Mock(Node)
+        def includingWithChildren = Mock(Node)
+        def root = "/some/root/"
+
+        hierarchy.recordNodeAccessingFileTree(excluding, root, excludes("child"))
+        hierarchy.recordNodeAccessingFileTree(including, root, includes("child"))
+        hierarchy.recordNodeAccessingFileTree(includingWithChildren, root, includes("child/**"))
+        expect:
+        assertNodesAccessing(root, excluding, including, includingWithChildren)
```

Comment:
I think this test fixture would benefit from separating the root from the assertions, for readability. Not to be done in this PR. Something like:
```
assertNodesAccessing(root) {
    hasNodes(excluding, including, includingWithChildren)
}
```

Reply:
👍